src/Mafia/Shell.hs

```diff
@@ -0,0 +1,373 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Mafia.Shell (
+    ShellError(..)
+  , renderShellError
+  , runShell
+  ) where
+
+import           Control.Monad.IO.Class (MonadIO(..))
+
+import qualified Data.Attoparsec.Text as Atto
+import qualified Data.Map as Map
+import qualified Data.Text as T
+
+import           Mafia.Hash
+import           Mafia.Home
+import           Mafia.IO
+import           Mafia.Path
+import           Mafia.Process
+
+import           P
+
+import           System.IO (IO)
+
+import           X.Control.Monad.Trans.Either (EitherT, hoistMaybe, hoistEither)
+
+
+data Script =
+  Script {
+      scriptId :: !Text
+    , scriptHash :: !Hash
+    , scriptPath :: !File
+    , scriptPackages :: ![Package]
+    , scriptSubmodules :: ![Submodule]
+    , scriptMain :: !Text
+    } deriving (Eq, Ord, Show)
+
+newtype Package =
+  Package {
+      unPackage :: Text
```

Comment:
I'm concerned this will lead to some fairly brittle scripts if it just jumps to the latest version each time. Would make it unsafe to use in prod (although I wouldn't be doing that, maybe it isn't a problem)